---
title: "Bayesian Network for Synthetic Data"
author: "Ted Laderas"
date: "5/4/2017"
output: html_document
---

In this file, I start generating the synthetic data using the gRain package.

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(gRain)
library(tidyverse)
library(data.table)
```

```{r functionblock}
byRow <- function(mat){
  rowVec <- lapply(1:nrow(mat), function(x){mat[x,]})
  outVec <- Reduce(c, rowVec)
  outVec
}

yn <- c("no","yes")


generateData <- function(jointTable, n){
  vec <- runif(n)
  
  #calculate cumulative probability for each category
  jointTable <- jointTable %>% mutate(cumProb = cumsum(prob))
  
  #lower bound for each table
  cumProbLow <- c(0, jointTable$cumProb)[1:nrow(jointTable)]
  jointTable <- data.frame(jointTable, cumProbLow)
  jointTable <- jointTable %>% filter(prob != 0)
  assign <- sapply(vec, function(x){ return(which(x < jointTable$cumProb & x > jointTable$cumProbLow))})
  
  dataOutFrame <- jointTable[assign,] %>% select(-cumProbLow, -prob, -cumProb)
#  dataOutFrame <- do.call(rbind, dataOut)
  return(dataOutFrame)
}


maleRiskScore <-function(df) {

  coeffList <- list(numAge=3.11296, numHtn=1.85508,   
                  numSmoke=0.70953, numT2D=0.5316, #numGenotype=0.55,
                  numBMI=.7927)
  
  coeffNames <- names(coeffList)
  
  len <- nrow(df)
  
  coeffList2 <- lapply(names(coeffList), function(x){vec <- 
    rep(coeffList[[x]], times=len)
    vec
  })
  
  names(coeffList2) <- names(coeffList)
  
  coeffFrame <- data.frame(coeffList2)
  
  coeffFrame <- coeffFrame %>% mutate(numHtn2 = 
                                        ifelse(df$treat=="Y", 1.99881, 
                                               1.93303)) %>% select(-numHtn, 
                                              numHtn=numHtn2)
  
  
  df2 <- df[,coeffNames]
  df2[,c("numAge", "numHtn", "numBMI")] <-   
    log(df2[,c("numAge", "numHtn", "numBMI")])
  coeffFrame <- coeffFrame[, coeffNames]
  
  coefSum <- rowSums(coeffFrame * df2)
  outProb <- (1 - 0.88431 ^ exp(coefSum - 23.988))
  
  return(data.frame(coefSum, outProb))
}



femaleRiskScore <-function(df) {

  coeffList <- list(numAge=2.72107, numHtn=2.81291,   
                  numSmoke=0.61868, numT2D=0.77763, #numGenotype=0.55,
                  numBMI=.51125)
  
  coeffNames <- names(coeffList)
  
  len <- nrow(df)
  
  coeffList2 <- lapply(names(coeffList), function(x){vec <- 
    rep(coeffList[[x]], times=len)
    vec
  })
  
  names(coeffList2) <- names(coeffList)
  
  coeffFrame <- data.frame(coeffList2)
  
  coeffFrame <- coeffFrame %>% mutate(numHtn2 = 
                                        ifelse(df$treat=="Y", 2.88267, 
                                               2.81291)) %>% select(-numHtn, 
                                              numHtn=numHtn2)
  
  
  df2 <- df[,coeffNames]
  df2[,c("numAge", "numHtn", "numBMI")] <-   
    log(df2[,c("numAge", "numHtn", "numBMI")])
  coeffFrame <- coeffFrame[, coeffNames]
  
  coefSum <- rowSums(coeffFrame * df2)
  outProb <- (1 - 0.94833 ^ exp(coefSum - 26.0145))
  
  return(data.frame(coefSum, outProb))
}

selsel <- function(x){sapply(as.character(x), selectInRange)}

convertYN <- function(x){sapply(x, function(y){
  ynVec <- c("N"=0,"Y"=1)
  return(ynVec[y])
})}


  selectInRange <- function(rangeString){
    outVal <- 0
    if(grepl("-",rangeString, fixed=TRUE)){
    rangeStrs <- strsplit(rangeString,split = "-")[[1]]
    rangeStrs <- as.numeric(rangeStrs)
    outVal <- round(runif(1, rangeStrs[1], rangeStrs[2]))
    
    }
    if(grepl("+",rangeString, fixed=TRUE)){
      outVal <- gsub("+","",rangeString, fixed=TRUE)
      #print(rangeString)
      ov <- as.numeric(outVal)
      outVal <- round(runif(1, ov, ov+5))
    }
    if(grepl("<", rangeString, fixed=TRUE)){
      outVal <- gsub("<","", rangeString, fixed=TRUE)
      ov <- as.numeric(outVal)
      outVal <- round(runif(1, ov-5, ov ))
    }
    outVal
  }
  
  # convertYN <- function(value){
  #   ynVec <- c("N"=0,"Y"=1)
  #   return(ynVec[value])
  # }
transformDataSet <- function(testData){

  testData <- testData[,!duplicated(colnames(testData))]

  newTestData <- testData %>% mutate(numAge=selsel(age),
                                   numBMI=selsel(bmi),
                                   numTchol=selsel(tchol),
                                   #numHtn=ifelse(convertYN(htn)==1, 
                                                 #rnorm(1, 180, 15),
                                                 #rnorm(1, 120,15)),
                                    #             180, 120),
                                   numTreat=convertYN(treat),
                                   numT2D=convertYN(t2d),
                                   numSmoke=convertYN(smoking),
                                   numGenotype=ifelse(genotype=="0001",
                                                      1,0))

    #add sbp value
    numHtn <- sapply(newTestData$htn, function(x){ifelse(x=="N",
                                                     floor(rnorm(1,120,10)),
                                                  floor(rnorm(1,180,10)))})
    
    #indTreat <- which(newTestData$treat == "Y")
    
    newTestData$numHtn <- numHtn

  malePatients <- newTestData %>% filter(gender=="M")
  femalePatients <- newTestData %>% filter(gender=="F")

  mp <- maleRiskScore(malePatients)
  fp <- femaleRiskScore(femalePatients)

  out <- rbind(data.frame(malePatients, mp), data.frame(femalePatients,fp))
#  out <- out %>% mutate(outProb = outScore/100)
  
  out
}

callCVD <- function(out){
    testVar <- runif(nrow(out))
    out <- data.frame(out, testVar)
    out <- out %>% mutate(cvd= ifelse(testVar < outProb, "Y", "N"))

}
  
generatePatientIDs <- function(data){
  patientID <- sample(0:1000000, nrow(data))
  patientID <- as.character(patientID)
  patientID <- sapply(patientID, function(x){
    numZeros <- 8 - nchar(x)
    zeroPad <- paste0(as.character(rep(0,numZeros)), collapse="")
    paste0("HHUID", zeroPad,x, collapse = "")
  })
  return(data.frame(patientID, data))
}

  
```

```{r}
cohort <- fread("~/Code/OHDSIqueries/data/cohorts_05022017.csv")
#recode these variables as categorical
cohort$outcome <- factor(cohort$outcome)
cohort$htn <- factor(cohort$htn)
cohort$htn_treatment <- factor(cohort$htn_treatment)
cohort$diabetes <- factor(cohort$diabetes)
#add age
cohort <- cohort %>% mutate(age=2017-year_of_birth)

cohort <- cohort[cohort$race != "Unknown",]
#cohort$race <- droplevels(cohort$race)

yn <- c("N","Y")


HTNtreat <- table(cohort$htn, cohort$htn_treatment)
condHTNtreat <- t(HTNtreat/rowSums(HTNtreat))
condHTNtreat <- t(matrix(nrow=2,c(100,0,40,60), byrow = TRUE, dimnames=list(yn,yn)))

out <- cut(cohort$age, breaks=c(0,20, 40, 55, 70,90))
cohort <- data.frame(cohort, ageCategory=out)

#cohort <- cohort %>% mutate(ageCategory = ntile(age,n=4))

tabOut <- table(out)
ageProb <- tabOut/sum(tabOut)
condHTNAge <- table(cohort$ageCategory, cohort$htn)
condHTNAgeprob <- t(condHTNAge/rowSums(condHTNAge))

tabTreatCVD <- table(cohort$outcome, cohort$htn_treatment)
condTreatCVD <- tabTreatCVD

tabCVDAge <- table(cohort$ageCategory, cohort$outcome)

#rownames(condCVDAge)


```


## Basic Synthetic Dataset

The first variable we add is Age. 

```{r cars}
yn <- c("N","Y")

##adjust age distribution
#ageProb <- c(0.07, 0.15, 0.24, 0.38, 0.16 )

ageLevels <- c("0-20","20-40", "40-55", "55-70","70-90")
a <- cptable(~age, values=ageProb*100, levels=ageLevels)

ageProb

condCVDAge <- t(matrix(nrow=5, ncol=2, 
                       data=c(97,3,97,3,91,9,75,25,70,35),byrow = TRUE,
                     dimnames = list(ageLevels, yn)))

a.c <- cptable(~cvd+age, values=condCVDAge, levels=yn)
a.htn <- cptable(~htn+age, values=condHTNAgeprob, levels=yn)
htn.treat <- cptable(~treat+htn, values=condHTNtreat, levels=yn)
treat.cvd <- cptable(~cvd+treat, values=c(50,50,10,90), levels=yn)

cvd.a.treat <- cptable(~cvd+treat+age, values=c(98,2, 0, 100,
                                                90,10, 0, 100,
                                                80,20, 0,100,
                                                100,0,0,100), levels=yn)
```

0 0 0-20
1 0 0-20
0 1 0-20
1 1 0-20
0 0 20-40
1 0 20-40
0 1 20-40
1 1 20-40
0 0 40-55
1 0 40-55
0 1 40-55
1 1 40-55
0 0 55-70
1 0 55-70

## Adding Smoking

Here I add smoking, which is associated with hypertension.

```{r}
a.s <- cptable(~smoking+age, values=c(95,5,90,10,85,15,75,25,90,10),levels=yn)
#s.c <- cptable(~cvd|smoking, values=c(40,60,10,90), levels=yn)
s.h <- cptable(~htn+smoking, values=c(50,50,33,67))

#plist <- compileCPT(list(a, a.c, a.htn, htn.treat, treat.cvd, s.c, a.s, s.h)) 
plist <- compileCPT(list(a, a.c, a.htn, htn.treat, treat.cvd, a.s, s.h)) 
net <- compile(grain(plist))

plot(net)

test <- querygrain(net, nodes=c("cvd","age", "htn","treat","smoking"), type="joint")

test <- data.frame(test, check.names = FALSE)

out <- data.frame(t(as.matrix(test)),check.names = FALSE)
out <- data.frame(cvdTreatHTNSmoke=row.names(out), out, check.names = FALSE)

testMelt <- gather(out, age, prob, -cvdTreatHTNSmoke) %>% filter(prob > 0) %>%
  separate(cvdTreatHTNSmoke, into=c("cvd", "treat", "htn","smoke"))

#testData <- generateData(testMelt, n=100000)

pRace <- table(cohort$race)/nrow(cohort)

names(pRace) <- c("AmInd", "Asian/PI", "Black/AfAm", "White")

race <- cptable(~race, values=pRace, levels=names(pRace))
#write.csv(testData, "~/Code/cvdWorkshop/night2CART/testData.csv",row.names = FALSE)
bmiLevels <- c("15-18","18-25", "25-31","31+")
bmiProb <- c(0.16, 0.68, 0.10, 0.06)

bmi <- cptable(~bmi, values=bmiProb, levels=bmiLevels)

valuesBMI <- c(97, 3, 98, 1, 98, 2, 99, 1,
            88, 12, 92,8, 92, 8, 96, 4,
            55, 45, 70, 30, 70, 30, 85, 15,
            20, 80, 33, 66, 33, 66, 65, 35)

##t2d.race.bmi is a joint cpt: p(t2d|race,bmi)
t2d.race.bmi <- cptable(~t2d|race:bmi, values=c(valuesBMI), levels=yn)
t2d.cvd <- cptable(~cvd|t2d, values=c(95, 5, 5, 95), levels= yn)
smoking <- cptable(~smoking, values=c(85, 15), levels=yn)
##tchol.genotype is a cpt 


plist <- compileCPT(list(a, a.c, a.htn, htn.treat, treat.cvd, 
                         smoking, s.h ,race, bmi, t2d.race.bmi, t2d.cvd)) 
net <- compile(grain(plist))

#testData <- simulate(net, 100000)

# table(testData$t2d)[2]/table(testData$t2d)[1]*100
# 
# test <- table(testData$bmi, testData$t2d)
# colSums(test)[2]/colSums(test)[1]

#write.csv(testData, "testData7vars.csv")

#test
```


## Selected SNPs

These first 2 snps are on the same chromosome and so I will model them as having identical risk, and having both does not increase risk. I am limiting myself to the homozygous variants (excluding the heterozygous variants) in order to simplify things.

[Rs10757278](http://snpedia.com/index.php/Rs10757278) (Chr 9, 22124478). (A;A) 0.78 x risk of heart disease, (A;G) 1.3 x risk, (G;G) 1.6x risk for heart disease. Co-morbidity: diabetes

[Rs1333049](http://snpedia.com/index.php/Rs1333049) (Chr 9, 22125504). (G;G) Normal (WT); (C;G) 1.5 increased risk; (C;C) 1.9x increased risk.

These last 2 snps are on different chromosomes for the first. Again, limiting to the homozygous cases to simplify things.

[Rs4665058](http://snpedia.com/index.php/Rs4665058), 4x risk if have two copies of the SNP (A;A), 2 X (A;C) and (C;C) - wild type. Chr 2, 159333698

[rs8055236](https://www.snpedia.com/index.php/Rs8055236) (Chr 16, 83178793) (T;T) Normal (WT);
(G;T) 1.9x risk, (G;G) 2.2x increased risk


```{r}
raceNames <- names(pRace)
#p("AA"|race) = c(21, 24, 80, 30)
#p("GG"|race) = c(79, 76, 20, 70)


snpLookup <- list("0000" = c("AA", "CC", "CC", "TT"),
     "0001" = c("AA", "CC", "CC", "GG"),
     "0010" = c("AA", "CC", "AA", "TT"),
     "1100" = c("GG", "GG", "CC", "TT"),
     "1110" = c("GG", "GG", "AA", "GG"),
     "1111" = c("GG", "GG", "AA", "GG"))

snpNames <- c("rs10757278", "rs1333049", "rs4665058", "rs8055236")

snpLookup <- lapply(snpLookup, function(x) {names(x) <- c("rs10757278", "rs1333049", "rs4665058", "rs8055236"); return(x)})



rs10757278race = matrix(nrow=2, byrow = TRUE, data=c(21, 24, 80, 30, 
                                                     79, 76, 20, 70), 
                        dimnames = list( c("AA","GG"),raceNames))/100

r1 <- rs10757278race

#p("AA"|race) = c(10, 10, 10, 10) 
#p("CC"|race) = c(80,80,80,80)

rs4665058race = 
  matrix(nrow=2, data= c(10, 10, 10, 10, 90,90,90,90), byrow=TRUE, 
         dimnames= list(c("AA","CC"), raceNames ))/100

r2 <- rs4665058race

#p("GG"|race) = c(62, 65, 20, 62)
#P("TT"|race) = c(38, 35, 80, 38)

rs8055236race = matrix(nrow=2, byrow=TRUE, data=c(62, 65, 20, 40, 
                                                  38, 35, 80, 60),
                       dimnames = list(c("GG","TT"), raceNames))/100
r3 <- rs8055236race

genotypes <- c("1111", "1110", "1100", "0010", "0001", "0000")

probList <- list()
probList[["0000"]] <- r1[2,] * r2[2,] * r3[2,]
probList[["1100"]] <- r1[1,] * r2[2,] * r3[2,]
probList[["1110"]] <- r1[1,] * r2[1,] * r3[2,]
probList[["0010"]] <- r1[2,] * r2[1,] * r3[2,]
probList[["0001"]] <- r1[2,] * r2[2,] * r3[1,]
probList[["1111"]] <- r1[1,] * r2[1,] * r3[1,]

cptgenoRace <- do.call(rbind,probList)
cptgenoRace <- cptgenoRace[genotypes,]

race.genotype <- cptable(~genotype|race, values=cptgenoRace, levels=genotypes, normalize=TRUE)

##tchol.cvd is a cpt: p(cvd|tchol)
tcholStates <- c("<160", "160-199", "200-239","240+")
tchol.cvd <- cptable(~cvd|tchol, values=c(100,0, 96,4, 88,12, 76,24), levels=yn)

gender <- cptable(~gender, c(45,55), levels=c("M","F"))

genotype.tchol <- cptable(~tchol|genotype, 
                          values = c(5,10,30,55,
                                     10,20,30,40,
                                     15,20,45,30,
                                     15,20,45,30,
                                     30,60,5,5,
                                     30,60,5,5), levels=tcholStates)

#c.a.t.d.tchol <- andtable(~cvd + age + treat, levels=yn)

plist <- compileCPT(list(a, a.htn, htn.treat, smoking, s.h, race,
                         bmi, t2d.race.bmi, race.genotype, 
                         genotype.tchol,
                         gender))

net <- compile(grain(plist))

testData <- simulate(net, nsim = 500000)
out <- transformDataSet(testData)
out <- generatePatientIDs(out)

out[out$genotype=="0001","outProb"] <- out[out$genotype=="0001","outProb"] + 0.07


out2 <- callCVD(out) 
cvdInds <- which(out2$cvd == "Y")
toRemove <- floor(length(cvdInds) * 0.6)
indRemove <- sample(cvdInds, toRemove)

out2 <- out2[-indRemove,]

table(out2$cvd)



gData <- out2 %>% filter(numAge < 45 & numAge >18)  

#gData[gData$genotype=="0001","outProb"] <- gData[gData$genotype=="0001","outProb"] + 0.10

write.csv(gData, "data/genoDataAllColumns.csv")

#gData2 <- callCVD(gData2)
#gData <- callCVD(gData)
#genotypeData2 <- callCVD(genotypeData)

table(gData$cvd)
#table(gData2$cvd)

gDataSmall <- gData[sample(1:nrow(gData), floor(nrow(gData)/3)),]

#pasteTable <- lapply(snpLookup, function(x){paste(x, collapse=",")})

genotypes <- lapply(as.character(gDataSmall$genotype), function(x){snpLookup[[x]]})

outMatrix <- matrix(ncol=4, nrow=length(genotypes), data="D")
colnames(outMatrix) <- snpNames

for(i in 1:length(genotypes)){
  #print(i)
  outMatrix[i,] <- genotypes[[i]]
}

genoFrame <- data.frame(outMatrix)
colnames(genoFrame) <- snpNames

save.image("syntheticdata.RData")

genoData <- data.frame(gDataSmall, genoFrame) 

write.csv(genoData, "../data/genoDataAllColumns.csv", row.names = FALSE)

#test <- gData %>% filter(numTchol < 200 & numBMI < 25)
#summary(test)
#table(test$genotype, test$cvd)

newData <- out2 %>% select(patientID, age, htn, treat, smoking, race, t2d, gender, numAge, bmi=numBMI, tchol=numTchol, sbp=numHtn,cvd)

write.csv(out2, "../data/fullPatientDataAllCols.csv", row.names=FALSE)

write.csv(newData, "../data/fullPatientData.csv", row.names = FALSE)

genoDataOut <- genoData %>% select(patientID, age, htn, treat, smoking, race, t2d, gender, numAge, bmi=numBMI, tchol=numTchol, sbp=numHtn, rs10757278, rs1333049, rs4665058,  rs8055236, cvd)

#make test and train sets for workshop

inds <- sample(1:nrow(genoDataOut), floor(.85 * nrow(genoDataOut)))
genoDataTrain <- genoDataOut[inds,]
genoDataTest <- genoDataOut[-inds,]

write.csv(genoDataTrain, "../data/genoDataTrainSet.csv", row.names=FALSE)
write.csv(genoDataTest, "../data/genoDataTestSet.csv", row.names=FALSE)

inds <- sample(1:nrow(newData), floor(.85 * nrow(newData)))
newDataTrain <- newData[inds,]
newDataTest <- newData[-inds,]

write.csv(newDataTrain, "../data/fullDataTrainSet.csv", row.names=FALSE)
write.csv(newDataTest, "../data/fullDataTestSet.csv", row.names=FALSE)



write.csv(genoDataOut, "~/Code/OHDSIqueries/data/genoData.csv", row.names = FALSE)



#newData <- data.frame(patientID, newData)


table(newData$cvd)
hist(out2$outProb)

out2$age <- ordered(out2$age)
out2$cvd <- factor(out2$cvd)
out2$htn <- factor(out2$htn)

library(party)
plot(ctree(formula = cvd ~ age + htn, data=out2))

summary(out)
```